### PR TITLE
interpreter: Support to_string method for int values

### PR DESCRIPTION
**Pull request by inigomartinez**
Although some other base types like boolean objects can be expresed as strings, this is not possible with int objects.

This patch adds support to express int values as strings as hex, decimal or octal values.

### CODE REVIEW

**Comment by nirbheek** [comment]
Fwiw, you can convert an `int` to a base 10 `string` representation using something like `'@0@'.format(42)`.

Somewhat relatedly: https://github.com/mesonbuild/meson/pull/2061, which is for `to_int()`.

File: mesonbuild/interpreterbase.py [hunk 1]
```diff
@@ -449,6 +449,16 @@ def int_method_call(self, obj, method_name, args):
...
             else:
                 raise InterpreterException('int.is_odd() must have no arguments.')
+        elif method_name == 'to_string':
+            if not posargs:
+                return str(obj)
+            elif len(posargs) == 1 and isinstance(posargs[0], str):
+                f = 'd' if len(posargs[0].strip()) == 0 else posargs[0]
+                if re.match('^[bcdoxX]$', f) is None:
+                    raise InvalidCode('Invalid format for int to string conversion "%s"' % f)
+                return str(('{:' + f + '}').format(obj))
```
Comment (nirbheek) [inline mesonbuild/interpreterbase.py]: Why not use a kwarg for the base instead? Wouldn't that be simpler to parse and to use?

**Comment by nirbheek** [comment]
To be honest, I can't actually think of a use-case for this that isn't also covered by `'@0@'.format()`. Do you need this for something? :)

**Comment by inigomartinez** [comment]
> Fwiw, you can convert an int to a base 10 string representation using something like '@0@'.format(42).

Yes, it's my approach at the moment, but it breaks out the consistency in the meson port I'm working on. Thank you for the suggestion :)

> Somewhat relatedly: #2061, which is for to_int().

Yes, very much related. It would be nice to have the two PR merged together.

**Comment by inigomartinez** [comment]
This is part of my current work:

```
output = '\nSystem paths:\n'
...
output += '\nMiscellaneous:\n'
output += '  have introspection: ' + enable_introspection.to_string() + '\n'
output += '  tests: ' + tests + '\n'
**output += '  more-asserts: @0@\n'.format(more_asserts)**
output += '  more-logging: ' + more_logging.to_string() + '\n'
output += '  valgrind: ' + valgrind_path + '\n'
output += '  code coverage: ' + get_option('b_coverage').to_string() + '\n'
output += '  LTO: ' + get_option('b_lto').to_string() + '\n'
output += '  Linker garbage collection: ' + enable_ld_gc.to_string() + '\n'
output += '  JSON validation for libnm: ' + enable_json_validation.to_string () + '\n'
output += '  sanitizers: ' + get_option('b_sanitize') + '\n'
output += '  Mozilla Public Suffix List: ' + enable_libpsl.to_string() + '\n'
message(output)
```
As you can see, all string options are _naturally_ joined, and _bool_ options are joined by using their `to_string` method, so it made sense for me to also have a `to_string` methods on _ints_.
 
And also, this is how I usually show a `configuration` output. I know that meson `configuration` tool has a prettier output, but this is shown just after meson ends parsing the files. I wanted to achieve something similar to what is done on autotools. Any suggestion to improve this is very appreciated.

File: mesonbuild/interpreterbase.py (hunk 1 above)
Comment (inigomartinez) [reply mesonbuild/interpreterbase.py]: I just followed [the approach](https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreterbase.py#L419) used by `bool`'s `to_string` method.

Although I don't have any preference, I just wanted to be consistent.

**Comment by nirbheek** [comment]
Since you're not using the base-specifier feature, it might be better to restrict this PR to base-10 string representations instead and skip the positional argument completely for now.

The difference is a quick merge because it's trivially useful vs several weeks/months of discussion about whether bases other than 10 are actually needed and whether we should use Python's syntax for them, and so on. :)

**Comment by nirbheek** [comment]
Also, a lot of people have asked about having a way to print a "configuration summary" instead of having to manually create it like you are doing. Perhaps you'd be interested in designing and implementing something like that too? Would be nice to have consistent summaries across all meson projects.

**Comment by inigomartinez** [comment]
Ok, I have reduced the  implementation to just convert the `int` value to base 10. I have also removed the test because it doesn't make any sense to have just one test.

Hope it's better now.

**Pull request merged**